consensus/src/runway/collection.rs

```diff
@@ -0,0 +1,536 @@
+    responses_from_network:
+        Receiver<UncheckedSigned<NewestUnitResponse<H, D, MK::Signature>, MK::Signature>>,
+    resolved_requests: Sender<Request<H>>,
+    collection: Collection<'a, MK>,
+}
+
+impl<'a, H: Hasher, D: Data, MK: KeyBox> IO<'a, H, D, MK> {
+    /// Create the IO instance for the specified collection and channels associated with it.
+    pub fn new(
+        round_for_creator: oneshot::Sender<Round>,
+        responses_from_network: Receiver<
+            UncheckedSigned<NewestUnitResponse<H, D, MK::Signature>, MK::Signature>,
+        >,
+        resolved_requests: Sender<Request<H>>,
+        collection: Collection<'a, MK>,
+    ) -> Self {
+        IO {
+            round_for_creator,
+            responses_from_network,
+            resolved_requests,
+            collection,
+        }
+    }
+
+    fn finish(self, round: Round) {
+        if self.round_for_creator.send(round).is_err() {
+            error!(target: "AlephBFT-runway", "unable to send starting round to creator");
+        }
+        if let Err(e) = self
+            .resolved_requests
+            .unbounded_send(Request::NewestUnit(self.collection.salt()))
+        {
+            warn!(target: "AlephBFT-runway", "unable to send resolved request:  {}", e);
+        }
+    }
+
+    /// Run the initial unit collection until it sends the initial round.
+    pub async fn run(mut self) {
+        use Status::*;
+        let mut catch_up_delay = futures_timer::Delay::new(Duration::from_secs(5)).fuse();
+        let mut delay_passed = false;
+
+        loop {
+            futures::select! {
+                response = self.responses_from_network.next() => {
+                    let response = match response {
+                        Some(response) => response,
+                        None => {
+                            warn!(target: "AlephBFT-runway", "Response channel closed.");
+                            return;
+                        }
+                    };
+                    match self.collection.on_newest_response(response) {
+                        Ok(Pending) => (),
+                        Ok(Ready(round)) => if delay_passed {
+                            self.finish(round);
+                            return;
+                        },
+                        Ok(Finished(round)) => {
+                            self.finish(round);
+                            return;
+                        },
+                        Err(e) => warn!(target: "AlephBFT-runway", "Received wrong newest unit response: {}", e),
+                    }
+                },
+                _ = catch_up_delay => match self.collection.status() {
+                    Pending => delay_passed = true,
+                    Ready(round) | Finished(round)  =>{
+                        self.finish(round);
+                        return;
+                    },
+                },
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
Maybe we could add a test for too few messages but some messages are sent more than once?

Reply:
Good idea, added.